### PR TITLE
[FluentCalendar] Use provided Culture for calendar operations

### PR DESCRIPTION
**Pull request by vrassouli**
# Pull Request

## 📖 Description
As reported in issue #1854, FluentCalendar (and FluentDatePicker) didn't use Culture for calendar operations and calculations.
It's now fixed, and the calendar now supports every culture correctly.

### 🎫 Issues
#1854 

### General

- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fluentui-blazor/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [ ] I have added [Unit Tests](https://github.com/Microsoft/fluentui-blazor/blob/master/unit-tests.md) for my new compontent
- [x] I have modified an existing component
- [x] I have validate [Unit Tests](https://github.com/Microsoft/fluentui-blazor/blob/master/unit-tests.md) for an existing component

### CODE REVIEW

File: examples/Demo/Shared/FluentUI.Demo.Shared.csproj [hunk 1]
```diff
@@ -58,6 +58,15 @@
 		<AdditionalFiles Include="$(ProjectDir)/*.xml" />
 	</ItemGroup>
 
+	<ItemGroup>
```
Comment (dvoituron) [inline examples/Demo/Shared/FluentUI.Demo.Shared.csproj:61]: You don't need to add these line. The csproj will do that automatically.

File: examples/Demo/Shared/Pages/DateTimes/Examples/DatePickerDefault.razor [hunk 2]
```diff
@@ -1,7 +1,7 @@
-<FluentDatePicker Label="Meeting date" @bind-Value="@SelectedValue" ReadOnly />
-<FluentDatePicker AriaLabel="To" @bind-Value="@SelectedValue" />
+﻿<FluentDatePicker Label="Meeting date" @bind-Value="@SelectedValue" ReadOnly />
+<FluentDatePicker AriaLabel="To" @bind-Value="@SelectedValue" Culture="@(new System.Globalization.CultureInfo("fa-IR"))" />
```
Comment (dvoituron) [inline examples/Demo/Shared/Pages/DateTimes/Examples/DatePickerDefault.razor:2]: Please, could you don't change the default sample. But add another Example if you want.

File: src/Core/Components/DateTime/CalendarExtended.cs [hunk 3]
```diff
@@ -159,36 +146,29 @@ public string GetYearsRangeLabel(int fromYear)
...
-    }
-
     /// <summary>
     /// Returns a list of days, abbreviated and complete (Mon, Monday), ...,(Sun, Sunday) in the correct culture.
     /// </summary>
     /// <returns></returns>
     public IEnumerable<(string Abbreviated, string Shorted, string Name)> GetDayNames()
     {
-        var firstDayOfWeek = (int)GetFirstDayOfWeek();
+        var firstDayOfWeek = Culture.DateTimeFormat.FirstDayOfWeek;
```
Comment (dvoituron) [inline src/Core/Components/DateTime/CalendarExtended.cs:171]: Could you replace this line by `(int)Culture.DateTimeFormat.FirstDayOfWeek` to keep the same "logic" and to remove the `(int)` from `return Shift(dayNames, (int)firstDayOfWeek);`

File: src/Core/Components/DateTime/CalendarExtended.cs [hunk 4]
```diff
@@ -41,46 +41,20 @@ public IEnumerable<DateTime> GetDaysOfWeek(int weekNumber, int monthOffset = 0)
...
+                yield return weekFirst.AddDays(i, Culture);
             }
         }
     }
 
-    /// <summary>
-    /// Gets the first day of the n-th month.
-    /// </summary>
-    /// <returns></returns>
-    public DateTime GetFirstDayToDisplay(int monthOffset = 0)
```
Comment (dvoituron) [inline src/Core/Components/DateTime/CalendarExtended.cs:61]: You cannot remove these `public` methods which are probably used by other developers.

File: src/Core/Components/DateTime/FluentCalendarDay.cs [hunk 5]
```diff
@@ -48,7 +48,7 @@ internal FluentCalendarDay(FluentCalendar calendar, DateTime day)
     /// <summary>
     /// Gets or sets a value indicating whether the day is selected by the user.
     /// </summary>
-    public bool IsSelected => Date.DayOfYear == _calendar.Value.GetValueOrDefault().DayOfYear;
+    public bool IsSelected => Date == _calendar.Value;
```
Comment (dvoituron) [inline src/Core/Components/DateTime/FluentCalendarDay.cs:51]: Why did you remove this code? Sometime, the "date" are same but not the "time" part. You need to include these cases also.

File: src/Core/Extensions/DateTimeExtensions.cs [hunk 6]
```diff
@@ -53,17 +65,20 @@ public static DateTime EndOfMonth(this DateTime self, CultureInfo culture)
     /// Returns the first day of the week.
     /// </summary>
     /// <param name="self"></param>
-    /// <param name="firstDayOfWeek"></param>
+    /// <param name="culture"></param>
     /// <returns></returns>
-    public static DateTime StartOfWeek(this DateTime self, DayOfWeek firstDayOfWeek)
+    public static DateTime StartOfWeek(this DateTime self, CultureInfo culture)
```
Comment (dvoituron) [inline src/Core/Extensions/DateTimeExtensions.cs:70]: You cannot remove the previous signature, maybe already used by many developers.

File: src/Core/Microsoft.FluentUI.AspNetCore.Components.csproj [hunk 7]
```diff
@@ -43,6 +43,9 @@
   <ItemGroup>
     <None Include="..\..\README.md" Pack="true" PackagePath="\" />
     <None Include="..\..\icon.png" Pack="true" PackagePath="\" />
+    <None Include="Components\DateTime\FluentCalendar.razor" />
```
Comment (dvoituron) [inline src/Core/Microsoft.FluentUI.AspNetCore.Components.csproj:46]: Like for the other csproj, these lines are not required

**Comment by vrassouli** [comment]
> @vrassouli please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
> 1. **Definitions**.
>    **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>    approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>    **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>    Project, including but not limited to communication on electronic mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>    associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>    described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work, You may Submit them separately
>    to the Project if You (a) retain all copyright and license information that was in the materials as You
>    received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>    containing materials of a third party:” followed by the names of the third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>    guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your employer has intellectual
>    property rights in Your Submission by contract or applicable law, You must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
> 
> * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>   Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>   the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>   parties.
> * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>   Your patent claims that are necessarily infringed by the Submission or the combination of the
>   Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>   import or otherwise dispose of the Submission alone or with the Project.
> * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>   No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>   granted by implication, exhaustion, estoppel or otherwise.
> 
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>    have disclosed under Section 3). You represent that You have secured permission from Your employer to
>    make the Submission in cases where Your Submission is made in the course of Your work for Your
>    employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>    have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>    You later become aware that would make Your representations in this Agreement inaccurate in any
>    respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>    information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>     supersedes any and all prior agreements, understandings or communications, written or oral, between
>     the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

@microsoft-github-policy-service agree

File: src/Core/Components/DateTime/CalendarExtended.cs (hunk 4 above)
Comment (vrassouli) [reply src/Core/Components/DateTime/CalendarExtended.cs:61]: Although the method is `public`, the containing class is `internal`.
And the method is not used anywhere in the project.
This seems to be redundant.

File: src/Core/Components/DateTime/FluentCalendarDay.cs (hunk 5 above)
Comment (vrassouli) [reply src/Core/Components/DateTime/FluentCalendarDay.cs:51]: Oops, you are right. I missed the "time" part.
But the DayOfYear is not correct too!
It returns the nth value of the day in a year.
This way, the DayOfYear for both '2023-01-01' and '2024-01-01' are the same, however they are not the same date!
I'll fix it

File: src/Core/Extensions/DateTimeExtensions.cs (hunk 6 above)
Comment (vrassouli) [reply src/Core/Extensions/DateTimeExtensions.cs:70]: May I add `[Obsolute]` attribute to it?

Comment (dvoituron) [reply src/Core/Extensions/DateTimeExtensions.cs:87]: Yes. It's for the other methods.

For public methods, you can't add `[Obsolete]` because in 95% of cases, it's a good value. You can create another signature with the `Culture`. But you have to manage both.

**Comment by vrassouli** [comment]
> I've already posted a few comments (first quick check).
> 
> I'm afraid of the number of "small" changes in the code base. E.g. `Year = year.GetDay(_calendar.Culture) == 1 && year.GetMonth(_calendar.Culture) == 1 ? year : year.StartOfYear(_calendar.Culture);`
> 
> Could you add unit tests to validate your changes and new features?

I've added some tests to ensure correct functionality by using Culture.

And one thing more, I noticed that `FluentCalendar_DayValues` of the `FluentCalendarTests` is wrong. Although it is passing, but it's not correct. It looks for `calendar.FindAll(".date")` and loops through the value to test the result. but the value contains nothing.

**Review by dvoituron — APPROVED** [review]
Great job. Thanks.
